src/Microsoft.ML.LightGbm/LightGbmArguments.cs

```diff
@@ -1,678 +1,351 @@
+        private protected OptionsBase BoosterOptions;
+    }
 
-        /// <summary>
-        /// Determines which booster to use.
-        /// </summary>
-        /// <value>
-        /// Available boosters are <see cref="DartBooster"/>, <see cref="GossBooster"/>, and <see cref="TreeBooster"/>.
-        /// </value>
-        [Argument(ArgumentType.Multiple, HelpText = "Which booster to use, can be gbtree, gblinear or dart. gbtree and dart use tree based model while gblinear uses linear function.", SortOrder = 3)]
-        public ISupportBoosterParameterFactory Booster = new TreeBooster.Options();
+    /// <summary>
+    /// Gradient boosting decision tree.
+    /// </summary>
+    /// <remarks>
+    /// For details, please see <a href="https://en.wikipedia.org/wiki/Gradient_boosting#Gradient_tree_boosting">gradient tree boosting</a>.
+    /// </remarks>
+    public sealed class GradientBooster : BoosterParameterBase
+    {
+        internal const string Name = "gbdt";
+        internal const string FriendlyName = "Tree Booster";
 
         /// <summary>
-        /// Determines whether to output progress status during training and evaluation.
+        /// The options for <see cref="GradientBooster"/>, used for setting <see cref="Booster"/>.
         /// </summary>
-        [Argument(ArgumentType.AtMostOnce, HelpText = "Verbose", ShortName = "v")]
-        public bool Verbose = false;
+        [TlcModule.Component(Name = Name, FriendlyName = FriendlyName, Desc = "Traditional Gradient Boosting Decision Tree.")]
+        public sealed class Options : OptionsBase
+        {
+            internal override BoosterParameterBase BuildOptions() => new GradientBooster(this);
+        }
 
-        /// <summary>
-        /// Controls the logging level in LighGBM.
-        /// </summary>
-        /// <value>
-        /// <see langword="true"/> means only output Fatal errors. <see langword="false"/> means output Fatal, Warning, and Info level messages.
-        /// </value>
-        [Argument(ArgumentType.AtMostOnce, HelpText = "Printing running messages.")]
-        public bool Silent = true;
+        public GradientBooster(Options options)
+        {
+            Contracts.CheckUserArg(options.MinimumSplitGain >= 0, nameof(Options.MinimumSplitGain), "must be >= 0.");
+            Contracts.CheckUserArg(options.MinimumChildWeight >= 0, nameof(Options.MinimumChildWeight), "must be >= 0.");
+            Contracts.CheckUserArg(options.SubsampleFraction > 0 && options.SubsampleFraction <= 1, nameof(Options.SubsampleFraction), "must be in (0,1].");
+            Contracts.CheckUserArg(options.FeatureFraction > 0 && options.FeatureFraction <= 1, nameof(Options.FeatureFraction), "must be in (0,1].");
+            Contracts.CheckUserArg(options.L2Regularization >= 0, nameof(Options.L2Regularization), "must be >= 0.");
+            Contracts.CheckUserArg(options.L1Regularization >= 0, nameof(Options.L1Regularization), "must be >= 0.");
```

Comment:
Don't you need this validation across all boosters?

Reply:
Yes - these are all in the BoosterParameterBase -- so GradientBooster, DartBooster and GossBooster all share these arguments. Therefore I added these restrictions. These restrictions come from the lightgbm documentation.

---
In reply to: [266665035](https://github.com/dotnet/machinelearning/pull/2948#discussion_r266665035) [](ancestors = 266665035)

Reply:
But you run validation only in `GradientBooster` `DartBooster` doesn't check this options as far as I can see.

---
In reply to: [266672503](https://github.com/dotnet/machinelearning/pull/2948#discussion_r266672503) [](ancestors = 266672503,266665035)

Reply:
You are right - I was thinking DartBooster derives from Gradient, but it doesnt (*it used to :))


---
In reply to: [266673967](https://github.com/dotnet/machinelearning/pull/2948#discussion_r266673967) [](ancestors = 266673967,266672503,266665035)